### ISSUE TITLE
Setting a block variable when the code isn't attached to the block header

### ISSUE THREAD

**Issue by rmillwood** [jmoenig/Snap#2422]
I often tinker in the block editor as I am trying out ideas and it seems strange that I can't work with the block variables unless my code is actualy attached to the block header:

![image](https://user-images.githubusercontent.com/2338141/60041912-61c39c80-96b4-11e9-8e16-f71614a151ba.png)

![image](https://user-images.githubusercontent.com/2338141/60042004-920b3b00-96b4-11e9-9e23-d84960d91c46.png)

Discovered this in Turtlestitch and then checked in Snap!

**Comment by rmillwood** [comment]
Sorry, the first image was supposed to show the pop-up without the 'a' variable available.

**Comment by jguille2** [comment]
Hi @rmillwood ,

Yes, you are right. But this is fine for me, and I think this is the right behavior.

The "blockEditor scripting area" is just like a patch of the whole "Sprite scripting area". The scope is the same. In this area you can use _global_ and _local_ (those of the same sprite) variables. You can use it in real time, test it... But "block variables" scope is just only that script definition: the blocks attached to the _hat_ of the block definition.

![blockVars](https://user-images.githubusercontent.com/2926315/60055021-8e8baa00-96dc-11e9-843a-c32d2754de2c.png)

To see it clearer, you can see that those "block variables" can't be used in that area, outside the block definition "chain".

![blockVar2](https://user-images.githubusercontent.com/2926315/60055273-5df84000-96dd-11e9-83b5-0dec29817042.png)

And really is the same case that the "scripts variables" in the general "scripting space".

![scriptVar](https://user-images.githubusercontent.com/2926315/60055134-e5917f00-96dc-11e9-969c-ab5cf6c6e848.png)
 
I think these "scopes" are fine a clear.

Joan

**Comment by brianharvey** [comment]
I disagree.  The Block Editor isn't just another piece of scripting area.  And people (including me!) often use the unattached space to build a subexpression that will later be attached to the script.  Also, someday (soonish, I hope) we are going to have internal definitions, which will make it even more important that the space around the script in a Block Editor be in the scope of the block.

**Comment by jmoenig** [comment]
Hmmm... I'm tending towards agreeing with Joan on this one. Currently, at least, what's inside a block editor doesn't have its own scope, so making block (-instance) variable in global custom blocks accessible in the same way we're making other script - scope variables accessible makes sense to me, and also visualizes scope conflicts correctly.

That said, Brian is right that in the future we want to extend the scope of block definitions to other local helper functions. But imo that's a slightly different issue than that of a *block instance* variable.

**Comment by brianharvey** [comment]
Oh.  I wasn't thinking of treating block variables differently from script variables.  I have a very simple idea here, namely that the block editor window has defining the block as its purpose, and the empty space in the window is just a convenient scratchpad for building the block.  So stuff in the scratchpad _will be_ in the script, and should have the scope of the script.  Same applies to formal parameters, upvars from called blocks, etc.

**Comment by jguille2** [comment]
Hi, only two notes...

- To avoid confusion (about this issue) I've just documented current behavior.

- You (both) have more much knowledge than I. But I see important difficulties in implementing Brian's proposal. Not because the scope issue, but because our live environment.
  - Script and Block variables are living inside their scripts. I can't test these reporters directly (clicking the reporter itself). If we change this (and "this" scripting area becomes the real "BlockScriptingArea"). What will be the result of the evaluation of this reporters (block variables)?
  - And the big problem is that the "block definition" is a prototype definition (not a single object). And then, which is its scope?

**Comment by rmillwood** [comment]
Joan said:
> "The "block editor scripting area" is just like a patch of the whole "sprite scripting area". The scope is the same. In this area you can use global and local (those of the same sprite) variables. You can use it in real time, test it..." 

I  have found that the 'block editor scripting area' is less functional than the 'sprite scripting area'. Exploring this has lead me to see that the 'green flag block', 'event blocks' and 'block definition blocks' (the last only ever seen in the 'block editor scripting area') are not imperative blocks, but simply declarations/headings/entry points. There may be a case for having tabs / independent windows for them all. BUT, by putting them in the 'sprite scripting area' you end up with your code for a sprite in one place (except your own blocks) and can use that space to set out things nicely to help with maintaining an overview.

All this came about because I failed to snap some code into the 'declaration block' in the 'block editor scripting area' leading to a block variable failing to be acknowledged, despite it being 'declared'. It took me a while to realise what was going and thanks to all commentators.

I would argue for a discussion about the designed structure of the Snap! language, [the different mental models needed](http://blog.richardmillwood.net/2018/01/31/mental-models-to-support-competence-in-computer-programming/), how the IDE currently supports them and how this might affect Snap!'s design - perhaps at the Snap! conference?

**Comment by brianharvey** [comment]
> perhaps at the Snap! conference?

Yeah!  Looking forward to it.

I may be wrong here, but I see this as another front in the ongoing struggle between a stance based on how the language is implemented and a stance based on the reasonable expectations of users.  I concede that my choice of language betrays which side I'm on, but this struggle predates Snap<i>!</i>.  Pretty much all implementations of Logo have had at least partial tail call elimination, since recursion is the main control structure.  (Logo is basically Lisp.)  Some versions of Logo have a TRACE command that logs entry to and return from selected procedures.  If you trace a tail-recursive procedure, should you see enter, enter, enter, return, return, return, or should you see enter, enter, enter, return?  The latter is the true picture of what the implementation does, but the former is the way we teach users to understand recursion.  Berkeley Logo chooses to disable tail call elimination in a procedure that's being traced.  It also goes through incredible contortions to get the error message "X didn't output to Y in Z" (where X, Y, and Z are procedures) to name the procedures that the user's mental model expects despite tail call elimination; unlike the TRACE example, we don't know ahead of time that this error will occur.

So, I think this is the same sort of issue.  Yes, there are implementation difficulties, but what the user expects is that everything in the block editor window is part of the block.
